### PR TITLE
[Profiler] Torch Profiler distributed info is not JSON serializable

### LINKED ISSUE #133308: Torch Profiler distributed info is not JSON serializable
### 🐛 Describe the bug

My code fails when trying to use torch profiler (https://github.com/pytorch/pytorch/blob/main/torch/profiler/profiler.py) to record memory traces in combination with torch distributed and Kineto, with the following error message:

Screenshot 2024-08-13 at 10 08 55

The function ` dist_info = self._get_distributed_info()`  returns a dictionary where one of the keys refer to 'pg_config'
```
        dist_info = {
            "backend": backend,
            "rank": dist.get_rank(),
            "world_size": dist.get_world_size(),
            "pg_count": dist.get_pg_count(),
            "pg_config": dist.distributed_c10d._get_all_pg_configs(),
        }
```

The problem is that the dictionary returned looks like:
`Distributed Info: {'backend': 'nccl', 'rank': 0, 'world_size': 1, 'pg_count': 2, 'pg_config': [{'pg_name': '0', 'pg_desc': 'default_pg', 'backend_config': 'cuda:nccl', 'pg_size': 1, 'ranks': [0]}, {'pg_name': '1', 'pg_desc': 'undefined', 'backend_config': 'cuda:nccl', 'pg_size': 1, 'ranks': [0]}], 'nccl_version': '2.21.5'}`
and hence it can't be serialised and the code fails in https://github.com/pytorch/pytorch/blob/9f17037e8ba7d98d8fbbcf40625fcad48c4c0b16/torch/profiler/profiler.py#L190
 
The error happens when there are more than 1 process groups, and then pg_config becomes a list of dictionaries.

### Versions

I am having this error with Pytorch 2.3 and 2.4 but the error does not happen when using 2.2 or lower 
And between those versions there was an update on the function
2.3 -> https://github.com/pytorch/pytorch/blob/97ff6cfd9c86c5c09d7ce775ab64ec5c99230f5d/torch/profiler/profiler.py#L279 
```
    def _get_distributed_info(self):
        import torch.distributed as dist

        if not dist.is_available() or not dist.is_initialized():
            return None

        backend = dist.get_backend()
        dist_info = {
            "backend": backend,
            "rank": dist.get_rank(),
            "world_size": dist.get_world_size(),
            "pg_count": dist.get_pg_count(),
            "pg_config": dist.distributed_c10d._get_all_pg_configs(),
        }
        if backend == "nccl":
            nccl_version = torch.cuda.nccl.version()
            dist_info["nccl_version"] = ".".join(str(v) for v in nccl_version)
        return dist_info

```
2.2.2 -> https://github.com/pytorch/pytorch/blob/39901f229520a5256505ec24782f716ee7ddc843/torch/profiler/profiler.py#L235
```
    def _get_distributed_info(self):
        import torch.distributed as dist

        if not dist.is_available() or not dist.is_initialized():
            return None

        return {
            "backend": dist.get_backend(),
            "rank": dist.get_rank(),
            "world_size": dist.get_world_size(),
        }
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @robieta @chaekit @aaronenyeshi @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise

### PR DESCRIPTION
**Pull request by sraikund16**
Summary: To fix https://github.com/pytorch/pytorch/issues/133308 we must create an encoder for numpy values so we can serialize the distributed metadata to JSON.

Test Plan: Added unit test to check that numpy values can be serialized

Differential Revision: D62411619

### CODE REVIEW

**Review by aaronenyeshi — APPROVED** [review]
LGTM, pleas check CI signals!

File: docs/source/conf.py [hunk 1]
```diff
@@ -2097,6 +2097,7 @@
     "NoneType",
     "NoopLogger",
     "NumberType",
+    "NumpyEncoder",
```
Comment (albanD) [inline docs/source/conf.py:2100]: Please do not add anything into this file.
You should either make this private (add an _ at the beginning) or properly document it.

**Review by albanD — CHANGES_REQUESTED** [review]

File: torch/profiler/profiler.py [hunk 2]
```diff
@@ -32,10 +32,33 @@
     "tensorboard_trace_handler",
     "profile",
     "ExecutionTraceObserver",
+    "NumpyEncoder",
```
Comment (albanD) [inline torch/profiler/profiler.py:35]: You can remove this entry now that it is not public anymore.

Comment (sraikund16) [reply torch/profiler/profiler.py]: removed!

**Review by albanD — APPROVED** [review]
SGTM from the public API point of view!

**Pull request closed without merging**